### ISSUE TITLE
Add proxy for doh requests

### ISSUE THREAD

**Issue by lhzw** [AdguardTeam/AdGuardHome#2096]
#1948 # Prerequisites

Please answer the following questions for yourself before submitting an issue. **YOU MAY DELETE THE PREREQUISITES SECTION.**

- [y] I am running the latest version
- [y] I checked the documentation and found no answer
- [y] I checked to make sure that this issue has not already been filed

### Problem Description

As I use DoH to avoid the dns pollution, but from mainland china, it's very slow to use https://1.1.1.1/dns-query or https://8.8.4.4/dns-query, or it's just blocked sometimes. So I'd like the doh requests can use the proxy, just like dnsmasq:
```
    environment:
      - http_proxy=socks5://192.168.8.188:3388
```

I tested the docker-compose environment variable, it doesn't work, either for the github update or doh requests. Then I put the proxy to guardhome's yaml config file, seems the github update uses it, but doh still doesn't use it.
It'll much faster and stable if doh can use a given proxy. Thank you very much.

### Proposed Solution

Let DoT and DoH use the given proxy, from the yaml config file or from the environment variable which is very flexible.

**Comment by ameshkov** [comment]
Well, if you want to use socks5, it would be better to extend this feature request. SOCKS5 proxy can be used to proxy DOT, and even plain DNS.

**Comment by lhzw** [comment]
It's just a example, I think this is better, from [wiki/configuration](https://github.com/AdguardTeam/AdGuardHome/wiki/Configuration):
http_proxy — proxy URL for HTTP client (e.g. "http://user:password@server:port/"). Supports "http", "https" and "socks5" scheme.

**Comment by ameshkov** [comment]
Yeah, that's right. But this proxy is not for upstreams, but for downloading filters, I suppose we'll need to add a different setting for proxying upstreams.

**Comment by Sil3ntVip3r** [comment]
AdGuard Home's `http_proxy` currently configures only its internal HTTP client.
DNS upstream construction does not receive proxy configuration, and current
dnsproxy `upstream.Options` has no general proxy or dialer injection point.

The smallest correct implementation order is:

1. Add an explicit proxy/dial abstraction to dnsproxy upstream options.
2. Cover DoH, DoT, and TCP DNS, and explicitly document whether UDP and DoQ are
   unsupported or require a relay.
3. Add a separate AdGuard Home upstream-proxy setting rather than silently
   reusing the filter-download `http_proxy`.
4. Pass it consistently to main, fallback, private, per-client, and
   upstream-test paths.
5. Validate configuration and redact credentials in logs and errors.

Related dnsproxy issues #62, #86, #336, and #410 should be linked here.
Existing dnsproxy PR #294 applies environment proxy behavior only to DoH and
does not satisfy this broader request.

Tests should use local fake CONNECT and SOCKS5 servers and cover transport
behavior, authentication, bootstrap policy, cancellation, unavailable
proxies, credential redaction, default direct behavior, and every upstream
category.
